include/gauge_path_helper.cuh

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include <gauge_field_order.h>
+#include <quda_matrix.h>
+#include <index_helper.cuh>
+#include <kernel.h>
+#include <thread_array.h>
+
+namespace quda {
+
+  template <int dim_>
+  struct paths {
+    static constexpr int dim = dim_;
+    const int num_paths;
+    const int max_length;
+    int *input_path[dim];
+    const int *length;
+    const double *path_coeff;
+    int *buffer;
+    int count;
+
+    paths(int ***input_path, int *length_h, double *path_coeff_h, int num_paths, int max_length) :
+      num_paths(num_paths),
+      max_length(max_length),
+      count(0)
+    {
+      // create path struct in a single allocation
+      size_t bytes = dim * num_paths * max_length * sizeof(int) + num_paths * sizeof(int);
+      int pad = ((sizeof(double) - bytes % sizeof(double)) % sizeof(double))/sizeof(int);
+      bytes += pad*sizeof(int) + num_paths*sizeof(double);
+
+      buffer = static_cast<int*>(pool_device_malloc(bytes));
+      int *path_h = static_cast<int*>(safe_malloc(bytes));
+      memset(path_h, 0, bytes);
+
+      for (int dir=0; dir<dim; dir++) {
+        // flatten the input_path array for copying to the device
+        for (int i = 0; i < num_paths; i++) {
+          for (int j = 0; j < length_h[i]; j++) {
+            path_h[dir * num_paths * max_length + i * max_length + j] = input_path[dir][i][j];
+            if (dir==0) count++;
+          }
+        }
+      }
+
+      // length array
+      memcpy(path_h + dim * num_paths * max_length, length_h, num_paths*sizeof(int));
+
+      // path_coeff array
+      memcpy(path_h + dim * num_paths * max_length + num_paths + pad, path_coeff_h, num_paths*sizeof(double));
+
+      qudaMemcpy(buffer, path_h, bytes, qudaMemcpyHostToDevice);
+      host_free(path_h);
+
+      // finally set the pointers to the correct offsets in the buffer
+      for (int d=0; d < dim; d++) this->input_path[d] = buffer + d*num_paths*max_length;
+      length = buffer + dim*num_paths*max_length;
+      path_coeff = reinterpret_cast<double*>(buffer + dim * num_paths * max_length + num_paths + pad);
+    }
+
+    void free() {
+      pool_device_free(buffer);
+    }
+  };
+
+  constexpr int flipDir(int dir) { return (7-dir); }
```

Comment:
I would love it if there was a future where we can move away from this MILC convention of encoding direction and dimension like this in a 1-d index and have a 2-d index instead, e.g., `[dim][dir]` like in the rest of QUDA.  Not necessarily for this PR, but what's your thoughts on how hard this would be to replumb?

Reply:
I think we could do something in the spirit of `std::pair<int,int>`, which would be a lot more readable. I do think it would be a lot of re-plumbing, though---there are a lot of arrays that look like `{ 0, 1, 7, 6 }` (but longer) that would need to become `{ {0, 1}, {1, 1}, {0, -1}, {1, -1} }` or something of the sort... the question is if the clarity added is worth the time spent debugging and propagating the changes out to the interface (and adding conversions to the MILC routines, etc)...